src/compiler/crystal/semantic/top_level_visitor.cr

```diff
@@ -884,6 +880,8 @@ class Crystal::TopLevelVisitor < Crystal::SemanticVisitor
     process_def_annotations(external, annotations) do |annotation_type, ann|
       if annotation_type == @program.call_convention_annotation
         call_convention = parse_call_convention(ann, call_convention)
+      elsif annotation_type == @program.primitive_annotation
```

Comment:
Would it be better to keep the rule that the [Primitive] annotation should have no method body and write the code like:

```crystal
{% if flag?(:interpreted) %} 
  @[Primitive(:interpreter_raise_without_backtrace)] 
  def raise_without_backtrace(exception : Exception) : NoReturn
{% else %}
  def raise_without_backtrace(exception : Exception) : NoReturn
    # ...
  end
{% end %}
```

Or is there other motivation to loose this restriction?

Reply:
The main reasons were:
- It's easier to just say "This method is a primitive in interpreter mode"
- Without this change, the entire body is now defined inside a macro, and so if an error happens the error trace will expand the macro. Error traces inside macros don't look as nice as regular traces.
- Primitive annotations are only used by us, the std/compiler developers. Others don't need to worry about this. And I think lifting this restrictions makes our lives a bit easier. It would be pretty strange to have a primitive with a body, a `Primitive` annotation, and for that to be unintended (like in this case)

Do you still think we should go back to the original behavior? Right now I don't have a preference.

Reply:
Let's keep this as is, it's fine. The "Error traces inside macros don't look as nice as regular traces" sold it for me.
And it's true that PrimitiveAnnotation is only aimed for internal usage so it's not really changing something for users.